skjson/src/FlattenType.sk

```diff
@@ -0,0 +1,211 @@
+/*****************************************************************************/
+/* Flattens the type structure into a set of paths associated to value. */
+/*****************************************************************************/
+
+module SKJSON;
+
+base class PrimType() uses Show {
+  children =
+  | PTInt()
+  | PTBool()
+  | PTFloat()
+  | PTString()
+  | PTEmptyArray()
+  | PTEmptyObject()
+
+  fun toString(): String
+  | PTInt() -> "int"
+  | PTBool() -> "bool"
+  | PTFloat() -> "float"
+  | PTString() -> "string"
+  | PTEmptyArray() -> "array"
+  | PTEmptyObject() -> "object"
+}
+
+fun flatten(
+  prefix: String,
+  iter: mutable Iterator<(String, CJSON)>,
+): mutable Iterator<(String, String, ?PrimType, ?Int, ?Float, ?String)> {
+  for (row in iter) {
+    (id, json) = row;
+    for (elt in flattenHelper(id, List[prefix], json)) {
+      yield elt
+    }
+  }
+}
+
+fun flattenHelper(
+  id: String,
+  path: List<String>,
+  json: CJSON,
+): mutable Iterator<(String, String, ?PrimType, ?Int, ?Float, ?String)> {
+  mkPath = () -> "".join(path.reversed());
+  json match {
+  | CJNull() -> yield (id, mkPath(), None(), None(), None(), None())
+  | CJBool(true) ->
+    yield (id, mkPath(), Some(PTBool()), Some(1), None(), None())
+  | CJBool(false) ->
+    yield (id, mkPath(), Some(PTBool()), Some(0), None(), None())
+  | CJInt(i) -> yield (id, mkPath(), Some(PTInt()), Some(i), None(), None())
+  | CJFloat(f) -> yield (id, mkPath(), Some(PTFloat()), None(), Some(f), None())
+  | CJString(s) ->
+    yield (id, mkPath(), Some(PTString()), None(), None(), Some(s))
+  | CJArray(arr) if (arr.isEmpty()) ->
+    yield (id, mkPath(), Some(PTEmptyArray()), None(), None(), None())
+  | CJArray(arr) ->
+    for (i => v in arr) {
+      for (elt in flattenHelper(id, List.Cons(".[" + i + "]", path), v)) {
```

Comment:
Can we change this to `List.Cons("[" + i + "]", path)`. i.e. remove the dot. This makes the path compatible with all the json tooling I've ever used and with JS too.

Reply:
Just to sanity check myself :D
```
> a = {b: [{c: 0}]}
{ b: [ { c: 0 } ] }
> a.b[0].c
0
> a.b.[0].c
a.b.[0].c
    ^

Uncaught SyntaxError: Unexpected token '['
```

Reply:
Ok will do!

Reply:
Done!